lib/fs/fakefs_test.go

```diff
@@ -172,3 +176,728 @@ func TestFakeFSRead(t *testing.T) {
+		t.Fatalf("got no error opening a non-existing file")
+	}
+
+	fd, err = fs.OpenFile("fOObar", os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	fd, err = fs.OpenFile("fOoBaR", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0664)
+	if err == nil {
+		fd.Close()
+		t.Fatalf("created an existing file while told not to")
+	}
+
+	fd, err = fs.OpenFile("FoObAr", os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	fd, err = fs.OpenFile("FOOBAR", os.O_RDWR, 0664)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+}
+
+func testFakeFSRemoveAll(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	filenames := []string{"bar", "baz", "qux"}
+	for _, filename := range filenames {
+		fd, err := fs.Create("/foo/" + filename)
+		if err != nil {
+			t.Fatalf("Could not create %s: %s", filename, err)
+		} else {
+			fd.Close()
+		}
+	}
+
+	if err := fs.RemoveAll("/foo"); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.Stat("/foo"); err == nil {
+		t.Errorf("this should be an error, as file doesn not exist anymore")
+	}
+
+	if err := fs.RemoveAll("/foo/bar"); err != nil {
+		t.Errorf("real systems don't return error here")
+	}
+}
+
+func testFakeFSRemoveAllInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	filenames := []string{"bar", "baz", "qux"}
+	for _, filename := range filenames {
+		fd, err := fs.Create("/FOO/" + filename)
+		if err != nil {
+			t.Fatalf("Could not create %s: %s", filename, err)
+		}
+		fd.Close()
+	}
+
+	if err := fs.RemoveAll("/fOo"); err != nil {
+		t.Errorf("Could not remove dir: %s", err)
+	}
+
+	if _, err := fs.Stat("/foo"); err == nil {
+		t.Errorf("this should be an error, as file doesn not exist anymore")
+	}
+
+	if err := fs.RemoveAll("/foO/bAr"); err != nil {
+		t.Errorf("real systems don't return error here")
+	}
+}
+
+func testFakeFSRemove(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/Foo/Bar")
+	if err != nil {
+		t.Fatal(err)
+	} else {
+		fd.Close()
+	}
+
+	if err := fs.Remove("/Foo"); err == nil {
+		t.Errorf("not empty, should give error")
+	}
+
+	if err := fs.Remove("/Foo/Bar"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Remove("/Foo"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testFakeFSRemoveInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/Foo/Bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	if err := fs.Remove("/FOO"); err == nil {
+		t.Errorf("not empty, should give error")
+	}
+
+	if err := fs.Remove("/Foo/BaR"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Remove("/FoO"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testFakeFSSameFile(t *testing.T, fs Filesystem) {
+	if runtime.GOOS == "windows" {
+		// windows time in not precise enough
+		t.SkipNow()
```

Comment:
Should do t.Skip("windows time in not precise enough")

Reply:
But hold on, this is not a real filesystem so the time can be as accurate as we want. On a real filesystem our modtime checks are not used, as we call os.SameFile, so not sure why this fails for fakefs on windows?

Reply:
I don't have a Windows box to test on, but I suppose it has something to do with time resolution beind worse on Windows than it is on other OSes in Go.  If I remember correctly, it used to be as bad as 15.6 ms before Go 1.7, then was supposedly fixed but still is nowhere near Unix-likes. I guess I could put a sleep into this test for completeness' sake.